pkg/reconciler/reconciler.go

```diff
@@ -0,0 +1,206 @@
+type InjectConfigFunc func(*rest.Config) error
+
+type reconciler struct {
+	finalizerName string
+	client        client.Client
+	recorder      record.EventRecorder
+	logger        *zap.Logger
+	EventingReconciler
+	filter       FilterFunc
+	finalizer    FinalizerFunc
+	injectConfig InjectConfigFunc
+	injectClient InjectClientFunc
+}
+
+var _ reconcile.Reconciler = &reconciler{}
+var _ inject.Client = &reconciler{}
+var _ inject.Config = &reconciler{}
+
+// inject.Client impl
+func (r *reconciler) InjectClient(c client.Client) error {
+	r.client = c
+	if r.injectClient != nil {
+		return r.injectClient(c)
+	}
+	return nil
+}
+
+// inject.Config impl
+func (r *reconciler) InjectConfig(c *rest.Config) error {
+	if r.injectConfig != nil {
+		return r.injectConfig(c)
+	}
+	return nil
+}
+
+// reconcile.Reconciler impl
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	obj := r.GetNewReconcileObject()
+	recObjTypeName := reflect.TypeOf(obj).Name() //TODO: Investigate why .Name() returns "". Need Name and not namespace.name
+
+	// TODO: Fix the code for cases where logger, recorder, or client are not set.
+	ctx := logging.WithLogger(context.TODO(), r.logger.With(zap.Any("request", request)))
+
+	if r.filter != nil && !r.filter(ctx, obj, r.recorder) {
+		logging.FromContext(ctx).Debug(fmt.Sprintf("Not reconciling %s as ShouldReconcile() returned false", recObjTypeName))
+		return reconcile.Result{}, nil
+	}
+
+	logging.FromContext(ctx).Debug(fmt.Sprintf("Reconciling %s", recObjTypeName))
+
+	if err := r.client.Get(ctx, request.NamespacedName, obj); err != nil {
+		if errors.IsNotFound(err) {
+			logging.FromContext(ctx).Error(fmt.Sprintf("Could not find %s", recObjTypeName))
+			return reconcile.Result{}, nil
+		}
+		logging.FromContext(ctx).Error(fmt.Sprintf("Error getting %s", recObjTypeName), zap.Error(err))
+		return reconcile.Result{}, err
+	}
+
+	if obj.GetDeletionTimestamp() != nil {
+		if r.finalizer != nil {
+			if err := r.finalizer(ctx, obj, r.recorder); err != nil {
+				logging.FromContext(ctx).Error("Finalizer func failed.", zap.Error(err))
+				return reconcile.Result{}, err
+			}
+		}
+		provisioners.RemoveFinalizer(obj, r.finalizerName)
+		return reconcile.Result{}, nil
+	}
+
+	if r.finalizerName != "" {
+		result := provisioners.AddFinalizer(obj, r.finalizerName)
+		if result == provisioners.FinalizerAdded {
+			if err := r.client.Update(ctx, obj); err != nil {
+				logging.FromContext(ctx).Error(
+					fmt.Sprintf("Error reconciling %s. Adding finalizer %s failed.", recObjTypeName, r.finalizerName),
+					zap.Error(err),
+				)
+				eventFailedReason := recObjTypeName + AddFinalizerFailed
+				r.recorder.Eventf(obj, corev1.EventTypeWarning, eventFailedReason, "%s reconciliation failed: %v", recObjTypeName, err)
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{Requeue: true}, nil
+		}
+	}
+
+	updateStatus, result, err := r.ReconcileResource(ctx, obj, r.recorder)
```

Comment:
My preference would be to calculate `updateStatus` here and not make every  `ReconcileResource` do it.